In [1]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder
from sklearn.svm import SVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn import datasets

### Reto Titanic

Carga el dataset del titanic, vamos a trabajar con este dataset.

In [2]:
titanic = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")
del(titanic["Name"])
del(titanic["PassengerId"])
del(titanic["Ticket"])
del(titanic["Cabin"])
titanic = titanic.dropna()

In [3]:
titanic["Embarked"].unique()

array(['S', 'C', 'Q'], dtype=object)

In [4]:
titanic["Pclass"].unique()

array([3, 1, 2], dtype=int64)

In [4]:
titanic["Embarked S"] = titanic["Embarked"].apply(lambda x:1 if x=='S' else 0)
titanic["Embarked C"] = titanic["Embarked"].apply(lambda x:1 if x=='C' else 0)
titanic["Embarked Q"] = titanic["Embarked"].apply(lambda x:1 if x=='Q' else 0)
titanic["Embarked N"] = titanic["Embarked"].apply(lambda x:1 if pd.isnull(x) else 0)
del(titanic["Embarked"])


titanic["Pclass 1"] = titanic["Pclass"].apply(lambda x:1 if x==1 else 0)
titanic["Pclass 2"] = titanic["Pclass"].apply(lambda x:1 if x==2 else 0)
titanic["Pclass 3"] = titanic["Pclass"].apply(lambda x:1 if x==3 else 0)
del(titanic["Pclass"])

titanic["Sex"] = titanic["Sex"].replace({'female': 1, 'male': 0})

In [6]:
titanic.to_csv('titanic_machine.csv', index=False)

In [6]:
titanic

Survived  Sex   Age  SibSp  Parch     Fare  Embarked S  Embarked C  \
0           0    0  22.0      1      0   7.2500           1           0   
1           1    1  38.0      1      0  71.2833           0           1   
2           1    1  26.0      0      0   7.9250           1           0   
3           1    1  35.0      1      0  53.1000           1           0   
4           0    0  35.0      0      0   8.0500           1           0   
..        ...  ...   ...    ...    ...      ...         ...         ...   
885         0    1  39.0      0      5  29.1250           0           0   
886         0    0  27.0      0      0  13.0000           1           0   
887         1    1  19.0      0      0  30.0000           1           0   
889         1    0  26.0      0      0  30.0000           0           1   
890         0    0  32.0      0      0   7.7500           0           0   

     Embarked Q  Embarked N  Pclass 1  Pclass 2  Pclass 3  
0             0           0         0         0         1  
1             0           0         1         0         0  
2             0           0         0         0         1  
3             0           0         1         0         0  
4             0           0         0         0         1  
..          ...         ...       ...       ...       ...  
885           1           0         0         0         1  
886           0           0         0         1         0  
887           0           0         1         0         0  
889           0           0         1         0         0  
890           1           0         0         0         1  

[712 rows x 13 columns]

Responde a las preguntas anteriores para este dataset
- ¿Que tipo de problema presenta?
        Datos de pasajeros, supervivientes o no, del titanic
- ¿Cuantas features hay?
        12 (originalmente 11)
- ¿Cuantas instancias hay?
        712 (originalmente 891)
- ¿Que tipo de etiqueta hay?
        Categorica: Survived
- Siguiendo la cheatsheet de sklearn (https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)
   ¿que algoritmo podriamos utilizar?
      SVC

- Despues realiza el split entre train y test.

In [7]:
X, y = titanic[titanic.columns[-12:]], titanic["Survived"]

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [9]:
X_train.shape

(569, 12)

In [10]:
X_test.shape

(143, 12)

- Escoge un modelo y entrenalo con estos datos. ¿Cual es su `score`? Para comprobar esto deberas leer la documentacion del modelo que hayas utilizado. Esto es algo que veremos mas en profundidad, pero intenta entenderlo y calcularlo por tu cuenta.


In [11]:
clf = make_pipeline(StandardScaler(), SVC(gamma='auto'))
clf.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(gamma='auto'))])

In [22]:
y_test_pred = clf.predict(X_test)
y_train_pred = clf.predict(X_train)

In [23]:
test_score = accuracy_score(y_test,y_test_pred)
test_score

0.8111888111888111

In [21]:
train_score = accuracy_score(y_train,y_train_pred)
train_score

0.8224956063268892

- Repite los dos pasos anteriores 10 veces en total. Utiliza una `seed` diferente cada vez que hagas el split, para obtener particiones del dataset distintas cada vez. Almacena los scores de train y test de cada split y calcula la media y la desviacion tipica de ambos y muestralos por pantalla.

In [30]:
scores = []
def simulacion_modelo(v):
    for i in range(v):
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)
        train_test_split(y, shuffle=False)
        clf = make_pipeline(StandardScaler(), SVC(gamma='auto'))
        clf.fit(X_train, y_train)
        y_pred = clf.predict(X_test)
        score_test = accuracy_score(y_test,y_pred)
        scores.append(score_test)
    return scores

simulacion_modelo(10)
scores = np.array(scores)
mean = np.mean(scores)
std = np.std(scores)

print(f'Media: {mean}\nDesviacion:  {std}')

Media: 0.8125874125874125
Desviacion:  0.02069741061111712
